### PR TITLE
Non-record: 24.7M params · int6 · Binary U-Net/SmearGate/BigramHash · 1.5hr · RTX 5060 Ti 16GB

### PR DESCRIPTION
**Pull request by randy06122001-boop**
## 24.7M params · int6 · Binary U-Net/SmearGate/BigramHash · 1.5hr · RTX 5060 Ti 16GB

**Non-record submission** - trained on RTX 5060 Ti 16GB for 1.5 hours.

### Approach
- Binary U-Net / SmearGate / BigramHash
- Quantized to 6 bits (int6)
- 24.7M parameters

Int6 Quantisation + 10L (5 Encoder - 5 Decoder) + Muon + 3x ReLU² MLP + SmearGate + BigramHash + SWA + RTX 5060 Ti

**val_bpb**: 1.4182 (roundtrip, seed=1337) | 11.63 MB artifact | NVIDIA RTX 5060 Ti, 2469 steps (~1.5h)
## Results (seed=1337, RTX 5060 Ti)

| Metric | Value |
| :--- | :--- |
| **val_bpb (Roundtrip)** | 1.4182 |
| val_loss | 2.3946 |
| Steps | 2469 |
| ms/step | 2187.4 |
| Training time | 5,400s (~1.5h) |
| Artifact | 11,633,008 bytes (11.63MB) |
| Parameters | 24,730,704 |

### Hardware & Environment
This run was executed on a local **NVIDIA GeForce RTX 5060 Ti (16GB)**. The environment uses Windows, so `torch.compile` (Triton) was disabled. The run demonstrates that advanced architectural features like SmearGate and Int6 quantization can significantly improve performance even on consumer-grade silicon.

## Architecture
- **10 transformer layers** (U-Net style: 5 encoder, 5 decoder blocks)
- **Model Dimension**: 512, 8 heads, 4 KV heads (GQA)
- **Quantization**: Int6 per-row for block weights, Int8/FP16 for others.
- **3x MLP expansion**: (hidden=1536) with ReLU² activation.
- **SmearGate**: causal blending of token embeddings with previous context.
- **BigramHash**: 4096-bucket hash embedding for consecutive token pairs.
- **ResidMix**: Learned residual blending across blocks.
- **Embedding**: Tied 1024-vocab embedding.

### Optimization
*   **Muon Optimizer**: Newton-Schulz iteration (5 steps) for all matrix parameters with 0.04 weight decay.
*   **AdamW**: Used for scalar parameters and embeddings with 0.04 weight decay.
*   **SWA (Stochastic Weight Averaging)**: Averaged 20 checkpoints during the warmdown phase for better generalization.
*   **Orthogonal Initialization**: Used for all matrix weights.

### Compression
*   **Int6 + Zstd-22**: Int6 quantization for block weights allows for significantly more parameters/layers within the 16MB limit, achieving a 3.73x payload compression ratio.

### Evaluation
*   **Roundtrip Validation**: Final metrics verified after dequantization to ensure performance persists in the exported artifact.

### Files
-  - quantized model
-  - training script
-  - training logs

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: 24.7M params · int6 · Binary U-Net/SmearGate/BigramHash · 1.5hr · RTX 5060 Ti 16GB

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

## Analysis ### PR Overview PR #997 by `randy` submits a 24.7M Int6 BinaryUNet model with SmearGate and BigramHash (1.5hr run, non-record 16MB track). Files changed: - `final_model.int8.ptz` (binary artifact) - `train_gpt.py` (training script) - `training_log.txt` ### Check 1: ILLEGAL n-gram family bug (target XOR'd into hash key) **CLEAN.** BigramHash is implemented at lines 697-700: ```python prev_ids = F.pad(token_ids[:, :-1], (1, 0), value=0) hash_ids = (prev_ids * 31 + token_ids) % self.num_buckets return self.proj(self.embed(hash_ids)) ``` The `token_ids` argument is `input_ids` (line 800: `self.bigram_hash(input_ids)`). The hash key uses only `(prev_token * 31 + current_token)` — both from the input sequence. `target_ids` are never passed to BigramHash nor XOR'd into the hash key. This is a legitimate bigram hash embedding over the input context. ### Check 2: ILLEGAL Pre-Quant TTT (multi-epoch on val_tokens without score-first) **NOT PRESENT.** There is no test-time training whatsoever in this script. `eval_val()` is called at lines 1130, 1238-1240, and 1289 — all under `model.eval()` + `torch.no_grad()`. No gradient steps are taken over `val_tokens` at any point. The warmup loop (lines 1086-1103) runs on `train_loader` data only and resets to initial weights afterward. ### Check 3: LEGAL score-first TTT (PR #1413 pattern, is_last_chunk guard) **NOT APPLICABLE.** No TTT of any kind is present. ### Check 4: HOLD scored-region SLOT **NOT APPLICABLE.** No scored-region slot manipulation detected. ### Conclusion This is a straightforward pure-neural training run. The BigramHash uses only input tokens in the key (no target leakage). Evaluation is inference-only with no gradient updates on validation data. The...

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
